### PR TITLE
Support an "expiration" date #726

### PR DESCRIPTION
**Pull request by g3wanghc**
First time using goLang and first time contribution. Please be gentle. :V

### CODE REVIEW

File: hugolib/site_test.go [hunk 1]
```diff
@@ -296,6 +296,42 @@ func TestDraftAndFutureRender(t *testing.T) {
 	viper.Set("BuildFuture", false)
 }
 
+func FutureExpirationRender(t *testing.T) {
```
Comment (bep) [inline hugolib/site_test.go]: I'm pretty sure it has to be named Test* something for it to run.

File: commands/hugo.go [hunk 2]
```diff
@@ -216,6 +217,7 @@ func initHugoBuildCommonFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&cleanDestination, "cleanDestinationDir", false, "Remove files from destination not found in static directories")
 	cmd.Flags().BoolVarP(&draft, "buildDrafts", "D", false, "include content marked as draft")
 	cmd.Flags().BoolVarP(&future, "buildFuture", "F", false, "include content with publishdate in the future")
+	cmd.Flags().BoolVarP(&expired, "buildExpired", "E", false, "include content with already expired")
```
Comment (bep) [inline commands/hugo.go]: Text: "include expired content"

File: docs/content/commands/hugo.md [hunk 3]
```diff
@@ -28,6 +28,7 @@ hugo
   -b, --baseURL string          hostname (and path) to the root, e.g. http://spf13.com/
   -D, --buildDrafts             include content marked as draft
   -F, --buildFuture             include content with publishdate in the future
+  -E, --buildExpired            include content already expired
```
Comment (bep) [inline docs/content/commands/hugo.md]: All content below /commands is autogenerated so no need to update these manually. Please remove these changes so this PR gets easier to read.

We usually do a ful update before a release.

File: docs/content/overview/configuration.md [hunk 4]
```diff
@@ -87,6 +87,8 @@ Following is a list of Hugo-defined variables that you can configure and their c
     buildDrafts:                false
     # include content with publishdate in the future
     buildFuture:                false
+    # includ content already expired
```
Comment (bep) [inline docs/content/overview/configuration.md]: include

File: hugolib/page.go [hunk 5]
```diff
@@ -486,6 +488,13 @@ func (p *Page) IsFuture() bool {
 	return true
 }
 
+func (p *Page) IsExpired() bool {
+	if p.ExpiryDate.After(time.Now()) {
```
Comment (bep) [inline hugolib/page.go]: This method should be a one-liner.

**Comment by g3wanghc** [comment]
@bep What's a good way to make 'make check' more verbose? I'm having trouble finding my formatting errors. 😢

File: hugolib/page.go [hunk 6]
```diff
@@ -467,7 +468,8 @@ func (p *Page) LinkTitle() string {
 }
 
 func (p *Page) ShouldBuild() bool {
```
Comment (bep) [inline hugolib/page.go]: This method has gotten complex enough to deserve a refactor.

I would pull out a function (as in not a method) with:

* buildFuture
* buildExpired
* buildDrafts
* publishDate
* expireDate

I.e. no dependency on Page nor Viper, then create a table driven test for that function.

**Comment by bep** [comment]
As to make and gofmt, the best is to actually do gfmting your code. But remember that make uses

`gofmt -s` (s = simplifies)

File: hugolib/site_test.go [hunk 7]
```diff
@@ -296,6 +296,42 @@ func TestDraftAndFutureRender(t *testing.T) {
...
+
+	s := siteSetup()
+
+	if len(s.Pages) > 1 {
+		t.Fatal("Expired content published unexpectedly")
+	}
+
+	if len(s.Pages) < 1 {
+		t.Fatal("Valid content expired unexpectedly")
+	}
```
Comment (bep) [inline hugolib/site_test.go]: These two assertions looks a little bit funky. What you want to know is that len(Pages) == 1 and s.Pages[0].title = "doc2"

Comment (g3wanghc) [inline hugolib/site_test.go]: Hahaha yeah my bad, I will fix that.

**Comment by g3wanghc** [comment]
I had to fix my commit messages hehehe. :V

File: hugolib/page.go (hunk 6 above)
Comment (g3wanghc) [inline hugolib/page.go]: @bep Is what I'm doing correct or does it need more work? 😁

**Comment by bep** [comment]
As to why this hasn't been merged. This looks perfectly fine, very good, and will be merged eventually. I have announced a kind-of freeze period before the 0.16 release, but that seems to drag out.

File: hugolib/page.go [hunk 8]
```diff
@@ -468,13 +468,46 @@ func (p *Page) LinkTitle() string {
...
+			return false
+		}
+	}
+	if !(buildDrafts || !Draft) {
+		return false
+	}
+	return true
+}
+
+func IsExpiredPage(expiryDate time.Time) bool {
```
Comment (moorereason) [inline hugolib/page.go]: The `IsExpiredPage` and `IsFuturePage` functions have nothing to do with pages, really.  They just compare times to `time.Now()`.  They should probably be named something like `IsExpiredTime` and `IsFutureTime`.

File: hugolib/page.go [hunk 9]
```diff
@@ -468,13 +468,46 @@ func (p *Page) LinkTitle() string {
...
+	if expiryDate.IsZero() {
```
Comment (moorereason) [inline hugolib/page.go]: Simplify this function and `IsFuturePage` to just be:

```go
return !expiryDate.IsZero() && !expiryDate.After(time.Now())
```

Or just get rid of these functions and roll that logic back into `AssertShouldBuild`.

File: hugolib/page.go [hunk 10]
```diff
@@ -468,13 +468,46 @@ func (p *Page) LinkTitle() string {
...
+			return false
 		}
 	}
-	return false
+	if !buildExpired {
+		if IsExpiredPage(expiryDate) {
+			return false
+		}
+	}
+	if !(buildDrafts || !Draft) {
```
Comment (moorereason) [inline hugolib/page.go]: We should test for drafts before we test for publish/expire dates.

**Comment by g3wanghc** [comment]
@moorereason 👌

File: hugolib/page.go [hunk 11]
```diff
@@ -468,13 +468,26 @@ func (p *Page) LinkTitle() string {
...
+	return AssertShouldBuild(viper.GetBool("BuildFuture"), viper.GetBool("BuildExpired"),
+		viper.GetBool("BuildDrafts"), p.Draft, p.PublishDate, p.ExpiryDate)
+}
+
+func AssertShouldBuild(buildFuture bool, buildExpired bool, buildDrafts bool, Draft bool,
+	publishDate time.Time, expiryDate time.Time) bool {
+	if !(buildDrafts || !Draft) {
+		return false
+	}
+	if !buildFuture {
```
Comment (moorereason) [inline hugolib/page.go]: Collapse these if-sets down one more level:

```go
if !buildFuture && !publishDate.IsZero() && publishDate.After(time.Now()) {
```

File: hugolib/page_test.go [hunk 12]
```diff
@@ -1089,3 +1089,49 @@ func compareObjects(a interface{}, b interface{}) bool {
...
+		{true, true, false, true, past, future, false},
+		{true, true, true, true, past, future, true},
+		{true, true, true, true, past, future, true},
+	}
+
+	for _, ps := range publishSettings {
+		s := AssertShouldBuild(ps.buildFuture, ps.buildExpired, ps.buildDrafts, ps.draft,
+			ps.publishDate, ps.expiryDate)
+		if s != ps.out {
+			t.Errorf("AssertShouldBuild unexpected output with prams: %+v", ps)
```
Comment (moorereason) [inline hugolib/page_test.go]: `prams` should be `params`?

Comment (g3wanghc) [inline hugolib/page_test.go]: Yes, yes it should be 😎

**Comment by bep** [comment]
This looks good. I took it for a spin, and the vital parts looks fine, but the `list expired` lists all of my pages -- I have only one with expiredDate.

**Comment by g3wanghc** [comment]
Oops, will fix it. :P

**Comment by g3wanghc** [comment]
@bep Fixed the list, and a typo. 

I've noticed something weird in `replacePage` on `site.go`.

```
func (s *Site) replacePage(page *Page) {
	// will find existing page that matches filepath and remove it
	s.removePage(page)
	s.addPage(page)
}
```

If the original page wasn't built, then it wouldn't be part of `site.Pages`. 

If you've modified an `unpublishedPage`, `replacePage` will be called but `removePage` wouldn't do anything since it can't find it in `site.Pages`. `addPage` will be called and `futureCount` is increased by `1`. 

If you added a future `publishDate` to an existing `Page` in `site.Pages`, then `removePage` will decrease `futureCount` by `-1`, and `addPage` will increate `futureCount` by `1`, resulting in a net gain of `0`.

Therefore, `Site stats` after `replacePage` is called will always be wrong.

Should we just keep track a list of `Pages` on `Site` that aren't being `built`?

**Comment by bep** [comment]
@g3wanghc create a separate issue for the stats counters.

**Comment by g3wanghc** [comment]
@bep Done. https://github.com/spf13/hugo/issues/2211

**Comment by bep** [comment]
Merged, solid work, thanks.

**Comment by g3wanghc** [comment]
Awesome!

**Pull request closed without merging**